truffle/src/main/java/org/jruby/truffle/core/string/StringNodes.java

```diff
@@ -1355,6 +1365,307 @@ private int prevCharHead(Encoding enc, byte[]bytes, int p, int s, int end) {
+                }
+                else if (repl != nil()) {
+                    final Rope replRope = rope((DynamicObject) repl);
+                    final ByteList replValue = RopeOperations.getByteListReadOnly(replRope);
+                    repBytes = replValue.unsafeBytes();
+                    rep = replValue.begin();
+                    replen = replValue.getRealSize();
+                    rep7bit_p = (replRope.getCodeRange() == CodeRange.CR_7BIT);
+                }
+                else if (encidx == UTF8Encoding.INSTANCE) {
+                    repBytes = SCRUB_REPL_UTF8;
+                    rep = 0;
+                    replen = repBytes.length;
+                    rep7bit_p = false;
+                }
+                else {
+                    repBytes = SCRUB_REPL_ASCII;
+                    rep = 0;
+                    replen = repBytes.length;
+                    rep7bit_p = false;
+                }
+                cr = CR_7BIT;
+
+                p = StringSupport.searchNonAscii(pBytes, p, e);
+                if (p == -1) {
+                    p = e;
+                }
+                while (p < e) {
+                    int ret = enc.length(pBytes, p, e);
+                    if (MBCLEN_NEEDMORE_P(ret)) {
+                        break;
+                    }
+                    else if (MBCLEN_CHARFOUND_P(ret)) {
+                        cr = CR_VALID;
+                        p += MBCLEN_CHARFOUND_LEN(ret);
+                    }
+                    else if (MBCLEN_INVALID_P(ret)) {
+                    /*
+                     * p1~p: valid ascii/multibyte chars
+                     * p ~e: invalid bytes + unknown bytes
+                     */
+                        int clen = enc.maxLength();
+                        if(buf == null){
+                            buf = RopeConstants.EMPTY_ASCII_8BIT_ROPE;
+                        }
+                        if (p > p1) {
+                               buf = makeConcatNode.executeMake(buf, RopeOperations.ropeFromByteList(new ByteList(pBytes, p1, p - p1)), enc);
+                        }
+
+                        if (e - p < clen){
+                            clen = e - p;
+                        }
+                        if (clen <= 2) {
+                            clen = 1;
+                        }
+                        else {
+                            int q = p;
+                            clen--;
+                            for (; clen > 1; clen--) {
+                                ret = enc.length(pBytes, q, q + clen);
+                                if (MBCLEN_NEEDMORE_P(ret)) break;
+                                if (MBCLEN_INVALID_P(ret)) continue;
+                            }
+                        }
+                        if (repBytes != null) {
+                            buf = makeConcatNode.executeMake(buf, RopeOperations.ropeFromByteList(new ByteList(repBytes, rep, replen)), enc);
```

Comment:
Similar deal here with not needing to create a `ByteList`. There's also a bit of a weird combination of slow & fast methods here. You probably should be using `MakeLeafRopeNode` rather than `RopeOperations` to create the leaf, unless you're on a slow path.